Runtime index on const array (of converted obj) causes C-compiler error

Indexing with a runtime index, into a const-array of const-objects (constructed via a converter), causes a C-compile error.

In the example below, _bar_ is a const array consisting of _b_, which is a const object constructed with the int-to-object converter _toObj_. Access into _bar_ with the runtime index _i_runtime_, causes a compile error in the C backend compiler.

### Example
<!-- Paste your example in the code-block below. -->
```
type    
    Obj = object    
        x: int    
    
    Tup = tuple    
        x: int    
    
converter toObj(x: int): Obj = Obj(x: x)
converter toTup(x: int): Tup = (x,)
                                   
const
    a = Obj(x: 0)    # uses normal object constructor
    b: Obj = 1       # uses object converter             
    b_tuple: Tup = 1 # uses tuple converter     
    foo = [a]                                  
    bar = [b]    
    bar_tuple = [b_tuple]
                             
let 
    bar_runtime = [b]
                         
proc main() =
    const i = 0  
    let i_runtime = 0
                         
    # works
    echo foo[i]
    echo bar[i]    

    # works                                                         
    echo foo[i_runtime]   

    # compile error in C-compiler
    echo bar[i_runtime]
           
    # works (with tuple) 
    echo bar_tuple[i_runtime] 
               
    # works (with runtime constructed array)
    echo bar_runtime[i_runtime]                                            

main() 
```

### Current Output
```
$ nim c -r const_array.nim
Hint: used config file '/home/zargot/projects/thirdparty/Nim/config/nim.cfg' [Conf]
Hint: used config file '/home/zargot/projects/thirdparty/Nim/config/config.nims' [Conf]
Hint: system [Processing]
Hint: const_array [Processing]
CC: const_array
Error: execution of an external compiler program 'gcc -c  -w  -I/home/zargot/projects/thirdparty/Nim/lib -I/home/zargot/projects/thirdparty/nim_bugs -o /home/zargot/.cache/nim/const_array_d/c
onst_array.c.o /home/zargot/.cache/nim/const_array_d/const_array.c' failed with exit code: 1

/home/zargot/.cache/nim/const_array_d/const_array.c:82:72: error: ‘T1_’ undeclared here (not in a function)
 NIM_CONST tyArray_eZVKehVP1XnECVJepajLtw bar_Iyn5d2bOFHV6BLKifeGnHw = {T1_}
                                                                        ^~~
```

### Additional Information
```
$ nim -v
Nim Compiler Version 0.19.9 [Linux: amd64]
Compiled at 2019-01-31
Copyright (c) 2006-2018 by Andreas Rumpf

git hash: ec6e5681daddde1ff81f9235a6f5348d46ab9eff
active boot switches: -d:release

$ gcc -v
Using built-in specs.
COLLECT_GCC=/usr/bin/gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/8.2.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++ --enable-shared --enable-threads=posix --enable-libmpx --with-system-zlib --with-isl --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-plugin --enable-install-libiberty --with-linker-hash-style=gnu --enable-gnu-indirect-function --enable-multilib --disable-werror --enable-checking=release --enable-default-pie --enable-default-ssp --enable-cet=auto
Thread model: posix
gcc version 8.2.1 20181127 (GCC)
```

Edit:
- cleanup
- add gcc version

--- a/changelog.md
+++ b/changelog.md
@@ -11,6 +11,8 @@
 
 ### Breaking changes in the compiler
 
+- Implicit conversions for `const` behave correctly now, meaning that code like `const SOMECONST = 0.int; procThatTakesInt32(SOMECONST)` will be illegal now.
+  Simply write `const SOMECONST = 0` instead.
 
 
 ## Library additions

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -595,6 +595,7 @@ proc semConst(c: PContext, n: PNode): PNode =
     if a.sons[length-2].kind != nkEmpty:
       typ = semTypeNode(c, a.sons[length-2], nil)
 
+    # don't evaluate here since the type compatibility check below may add a converter
     var def = semExprWithType(c, a[^1])
     if def.typ.kind == tyProc and def.kind == nkSym:
       if def.sym.kind == skMacro:

--- a/compiler/semstmts.nim
+++ b/compiler/semstmts.nim
@@ -595,11 +595,7 @@ proc semConst(c: PContext, n: PNode): PNode =
     if a.sons[length-2].kind != nkEmpty:
       typ = semTypeNode(c, a.sons[length-2], nil)
 
-    var def = semConstExpr(c, a.sons[length-1])
-    if def == nil:
-      localError(c.config, a.sons[length-1].info, errConstExprExpected)
-      continue
-
+    var def = semExprWithType(c, a[^1])
     if def.typ.kind == tyProc and def.kind == nkSym:
       if def.sym.kind == skMacro:
         localError(c.config, def.info, errCannotAssignMacroSymbol % "constant")
@@ -621,7 +617,10 @@ proc semConst(c: PContext, n: PNode): PNode =
         def = fitRemoveHiddenConv(c, typ, def)
     else:
       typ = def.typ
-    if typ == nil:
+
+    # evaluate the node
+    def = semConstExpr(c, def)
+    if def == nil:
       localError(c.config, a.sons[length-1].info, errConstExprExpected)
       continue
     if typeAllowed(typ, skConst) != nil and def.kind != nkNilLit:
@@ -639,7 +638,7 @@ proc semConst(c: PContext, n: PNode): PNode =
       b.sons[length-2] = a.sons[length-2]
       b.sons[length-1] = def
 
-    for j in 0 .. length-3:
+    for j in 0 ..< length-2:
       var v = semIdentDef(c, a.sons[j], skConst)
       if sfGenSym notin v.flags: addInterfaceDecl(c, v)
       elif v.owner == nil: v.owner = getCurrOwner(c)

--- a/tests/vm/teval1.nim
+++ b/tests/vm/teval1.nim
@@ -25,3 +25,18 @@ doAssert x == ""
 static:
   var i, j: set[int8] = {}
   var k = i + j
+
+type
+  Obj = object
+    x: int
+
+converter toObj(x: int): Obj = Obj(x: x)
+
+# bug #10514
+block:
+  const
+    b: Obj = 42
+    bar = [b]
+
+  let i_runtime = 0
+  doAssert bar[i_runtime] == b

